### PR TITLE
Hot Reload: Handle file deletions correctly

### PR DESCRIPTION
**Pull request by tmat**
Fixes file watcher and workspace updating to support file deletes.

Fixes https://github.com/dotnet/sdk/issues/42535

### CODE REVIEW

**Comment by tmat** [comment]
@phil-allen-msft ptal

File: src/BuiltInTools/DotNetDeltaApplier/StartupHook.cs [hunk 1]
```diff
@@ -26,9 +26,11 @@ public static void Initialize()
         // When launching the application process dotnet-watch sets Hot Reload environment variables via CLI environment directives (dotnet [env:X=Y] run).
         // Currently, the CLI parser sets the env variables to the dotnet.exe process itself, rather then to the target process.
         // This may cause the dotnet.exe process to connect to the named pipe and break it for the target process.
-        if (Path.ChangeExtension(processPath, ".exe") != Path.ChangeExtension(s_targetProcessPath, ".exe"))
+        var processExe = Path.ChangeExtension(processPath, ".exe");
```
Comment (phil-allen-msft) [inline src/BuiltInTools/DotNetDeltaApplier/StartupHook.cs:29]: I don't think the extension is guaranteed on Linux.  What would you expect to happen in that case?

File: src/BuiltInTools/dotnet-watch/HotReload/IncrementalMSBuildWorkspace.cs [hunk 2]
```diff
@@ -15,6 +15,7 @@
 using Microsoft.CodeAnalysis.ExternalAccess.Watch.Api;
 using System.Collections.Immutable;
 using Microsoft.CodeAnalysis.Text;
+using Microsoft.DotNet.Watcher.Internal;
```
Comment (phil-allen-msft) [inline src/BuiltInTools/dotnet-watch/HotReload/IncrementalMSBuildWorkspace.cs:18]: Are these usings sorted the way you want?

File: src/BuiltInTools/dotnet-watch/HotReloadDotNetWatcher.cs [hunk 3]
```diff
@@ -336,24 +336,43 @@ await Task.WhenAll(
...
+            static string GetSingularMessage(ChangeKind kind)
+                => kind switch
+                {
+                    ChangeKind.Update => "File updated",
+                    ChangeKind.Add => "File added",
+                    ChangeKind.Delete => "File deleted",
+                    _ => throw new InvalidOperationException()
+                };
+
+            static string GetPlurarMessage(ChangeKind kind)
```
Comment (phil-allen-msft) [inline src/BuiltInTools/dotnet-watch/HotReloadDotNetWatcher.cs:368]: (nit) do you mean GetPluralMessage?

File: src/BuiltInTools/dotnet-watch/HotReloadDotNetWatcher.cs [hunk 4]
```diff
@@ -336,24 +336,43 @@ await Task.WhenAll(
...
+                => kind switch
+                {
+                    ChangeKind.Update => "Files updated",
```
Comment (phil-allen-msft) [inline src/BuiltInTools/dotnet-watch/HotReloadDotNetWatcher.cs:371]: Should these "File updated", "File added" be translatable strings?

File: src/BuiltInTools/dotnet-watch/Internal/FileWatcher/ChangeKind.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.DotNet.Watcher.Internal;
+
+internal enum ChangeKind
+{
+    Update,
```
Comment (phil-allen-msft) [inline src/BuiltInTools/dotnet-watch/Internal/FileWatcher/ChangeKind.cs:8]: I sometimes see code that is sensitive to "Move" as being different from "Update".  Is there a case for that here?

File: src/BuiltInTools/DotNetDeltaApplier/StartupHook.cs (hunk 1 above)
Comment (tmat) [reply src/BuiltInTools/DotNetDeltaApplier/StartupHook.cs:29]: If the path has no extension `.exe` is appended.

Comment (tmat) [reply src/BuiltInTools/DotNetDeltaApplier/StartupHook.cs:29]:
```
> Path.ChangeExtension(@"/a/b", ".exe")
"/a/b.exe"
> Path.ChangeExtension(@"/a/b.exe", ".exe")
"/a/b.exe"
> Path.ChangeExtension(@"/a/b.dll", ".exe")
"/a/b.exe"
```

File: src/BuiltInTools/dotnet-watch/HotReloadDotNetWatcher.cs (hunk 4 above)
Comment (tmat) [reply src/BuiltInTools/dotnet-watch/HotReloadDotNetWatcher.cs:371]: Eventually, yes. Watch has many non-translatable strings. We should do a pass to move all strings to resources.

File: src/BuiltInTools/dotnet-watch/Internal/FileWatcher/ChangeKind.cs (hunk 5 above)
Comment (tmat) [reply src/BuiltInTools/dotnet-watch/Internal/FileWatcher/ChangeKind.cs:8]: Not sure what you mean. I'm guessing you're suggesting adding "Move" ChangeKind.
I think that might need to be done to address https://github.com/dotnet/sdk/issues/43320.

File: src/BuiltInTools/dotnet-watch/HotReloadDotNetWatcher.cs (hunk 3 above)
Comment (tmat) [reply src/BuiltInTools/dotnet-watch/HotReloadDotNetWatcher.cs:368]: yup

**Review by phil-allen-msft — APPROVED** [review]

**Pull request merged**